### DISCUSSION TITLE
Using Sioyek as a pdf viewer in Emacs

### DISCUSSION THREAD

**Discussion by ejuarezg** [ahrm/sioyek#347] [Q&A] [answered]
Has anyone figured out the configuration to get Sioyek working as a PDF viewer in Emacs? I'm (very) slowly learning Emacs and have only managed to get the following code working
```lisp
(setq TeX-view-program-list '(("Sioyek" "sioyek %o")))
(setq TeX-view-program-selection '((output-pdf "Sioyek")))
(setq +latex-viewers '(sioyek))
```
However, I haven't figured out how to get forward and inverse synctex search working.

NB: I'm using the `latex` module in Doom Emacs for an out-of-the-box working TeX configuration.

**Comment by JHonaker** [comment, marked as answer]
In order to setup Sioyek for SyncTeX with Emacs, you'll need to modify a few customization options.

First, you will need to add Sioyek to the `TeX-view-program-list` like you have, but there are a few more things that need to go there to get SyncTeX working. You also have to replace the default PDF viewer in `TeX-view-program-selection` which determines which program to use based on a series of predicates. For PDFs, you'll want to change the one listed under `output-pdf`.

```elisp
(custom-set-variables
  '(TeX-view-program-list
    '(("Sioyek"
      ("sioyek %o --reuse-instance"
        (mode-io-correlate " --forward-search-file %b --forward-search-line %n --inverse-search \"emacsclient --no-wait +%2:%3 %1\""))
       "sioyek")))
 '(TeX-view-program-selection
   '(((output-dvi has-no-display-manager)
      "dvi2tty")
     ((output-dvi style-pstricks)
      "dvips and gv")
     (output-dvi "xdvi")
     (output-pdf "Sioyek") ;; Here. By default this was "Evince" on my machine.
     (output-html "xdg-open")))
```

This allows you to refer to the "Sioyek" viewer like you would the built-in viewers. `mode-io-correlate` is a predicate defined in AucTeX that is only active when you have SyncTeX turned on.

Just for reference, the full command would read

    sioyek %o --reuse-instance --forward-search-file %b --forward-search-line %n --inverse-search "emacsclient --no-wait +%2:%3 %1"

What each of the substitutions and subcommands are:
- `%o` - The output pdf
- `--reuse-instance` - Does what it says on the tin. Reuse the instance instead of opening a new one.
- `--forward-search-file %b` - Search in `%b` which expands into the TeX filename used by SyncTeX
- `--forward-search-line %n` - Go to the current line number in the buffer expanded from `%n`
- `--inverse-search "emacsclient --no-wait +%2:%3 %1"` - Gives Sioyek the command to call when you right click with SyncTeX enabled (F4) which is run open `emacsclient` to the buffer for file `%1`, line `%2`, column `%3`.

To enable SyncTeX you'll need to turn on `TeX-source-correlate-mode`. Here's how you can do that with a hook that starts when a LaTeX file is loaded. The keybinding allows you to control-click on a place in the buffer, and have it jumped to in Sioyek.

```elisp
(add-hook 'LaTeX-mode-hook #'TeX-source-correlate-mode)
(with-eval-after-load 'latex
  (define-key TeX-source-correlate-map [C-down-mouse-1] #'TeX-view-mouse))
```

**Reply by ejuarezg** [reply]
This works beautifully! Thank you for the detailed explanation. For anyone interested, here are the minor changes I had to make to fit it into my Doom Emacs config

```lisp
(setq TeX-view-program-list
      '(("Sioyek"
        ("sioyek %o --reuse-instance"
         (mode-io-correlate " --forward-search-file %b --forward-search-line %n --inverse-search \"emacsclient --no-wait +%2:%3 %1\""))
        "sioyek"))
      TeX-view-program-selection
      '(((output-dvi has-no-display-manager)
         "dvi2tty")
        ((output-dvi style-pstricks)
         "dvips and gv")
        (output-dvi "xdvi")
        (output-pdf "Sioyek")
        (output-html "xdg-open"))
      +latex-viewers '(sioyek))

(add-hook 'LaTeX-mode-hook #'TeX-source-correlate-mode)

(with-eval-after-load 'latex
  (define-key TeX-source-correlate-map [C-down-mouse-1] #'TeX-view-mouse))
```

Without the `+latex-viewers '(sioyek)`, Emacs would still open the PDF in Evince.

**Reply by JHonaker** [reply]
Ah yea, that `+latex-viewers` is a Doom specific thing. I don't use Doom, so I have no idea what it does.

**Reply by eshrh** [reply]
For anybody else who comes across this issue, on newer auctex versions (Jun 20 2022) it's no longer necessary to add sioyek to `TeX-view-program-list`, I had sent a patch to add it to the builtins.

You only need to set `TeX-view-program-selection` to include `output-pdf "Sioyek"`.

**Reply by AlexD97** [reply]
Has anyone else the problem that sioyek doesn't show anything when auctex compiles the pdf? I have to manually reload the file in sioyek.

**Reply by Ziqi-Yang** [reply]
in `TeX-view-program-list`, don't add `--reuse-instance` argument to ensure showing

**Reply by AlexD97** [reply]
Somehow I don't encounter the problem anymore. I don't know why this happened.

**Reply by EssenSea** [reply]
> For anybody else who comes across this issue, on newer auctex versions (Jun 20 2022) it's no longer necessary to add sioyek to `TeX-view-program-list`, I had sent a patch to add it to the builtins.
> 
> You only need to set `TeX-view-program-selection` to include `output-pdf "Sioyek"`.

``` lisp
(use-package tex
  ...
:config
...
(setq TeX-view-program-seleciton '((output-pdf "Sioyek")))
...
```
the enviroment is gentoo linux, emacs 30.1, and while I was pressing C-c C-v trying to open pdf after compiler finished its work, emacs give back to me:

Wrong type argument: stringp, nil

Is there something wrong?

**Comment by idlip** [comment]
Just bumping it here. There is a emacs package in development which aims to replace doc-view and pdf-tools

Looks promising, and almost comparable to sioyek.

https://codeberg.org/divyaranjan/emacs-reader/
